Implement operation to simplify negated join on CsgTree

esseivaju opened the pull request:
Required to generate infix notation for  #1286  as part of #1282. The goal is to eliminate negated `Join` operations as those cannot be represented with infix notation.

esseivaju committed:
basic infix evaluator

esseivaju committed:
add test

esseivaju committed:
check for bounds

esseivaju committed:
address part of the comments

esseivaju committed:
check bounds with CELER_ASSUME

esseivaju committed:
address review

esseivaju committed:
take functor to evaluate sense

esseivaju committed:
doc

esseivaju committed:
add a method to simplify negated join in CsgTree

esseivaju committed:
if-init

sethrj commented on src/orange/orangeinp/CsgTree.cc:
```diff
@@ -166,6 +166,65 @@ auto CsgTree::simplify(NodeId node_id) -> Simplification
+ * This is required if the tree's logic expression is used with
+ * \c InfixEvaluator as negated joins are not supported.
+ */
+void CsgTree::simplify_negated_joins()
```
Is it possible to make this an external `CsgTreeUtils` function rather than extending the class?

esseivaju committed:
moved transform_negated_joins to CsgTreeUtils

esseivaju committed:
Merge branch 'develop' into simplify-negated-join

esseivaju committed:
Merge branch 'develop' into simplify-negated-join

esseivaju committed:
implement transformation following DeMorgan's law

esseivaju commented:
@sethrj @elliottbiondo I still need to add tests and finalize the implementation but I think the logic is good now.

esseivaju committed:
update node_id of childrens

esseivaju committed:
Correctly handle double negation of joined operands. Fix erase before lookup

esseivaju committed:
add simple test case

esseivaju commented:
I'll add more complex test cases but this is ready for an initial review

esseivaju committed:
fix node_id lookup when replacing Joined node

esseivaju committed:
use std::transform

esseivaju committed:
add more complex test case and fix issue when adding an unmodified join expression

esseivaju committed:
more test cases

esseivaju committed:
unmark children orphan Negated nodes when a Joined is unmarked as orphan and add test case

esseivaju committed:
yet another test case

esseivaju committed:
more fix and test

sethrj commented on src/orange/orangeinp/CsgTreeUtils.cc:
```diff
@@ -165,6 +171,273 @@ void simplify(CsgTree* tree, NodeId start)
+ * This is required if the tree's logic expression is used with
+ * \c InfixEvaluator as negated joins are not supported.
+ */
+CsgTree transform_negated_joins(CsgTree const& tree)
```
This function is too long; is there a sensible way to break it up into a detail helper class? Probably making these maps as class data, maybe a couple helper functions that you could call in sequence, turn the lambda into a member function.

sethrj commented on src/orange/orangeinp/CsgTreeUtils.cc:
```diff
@@ -165,6 +171,273 @@ void simplify(CsgTree* tree, NodeId start)
+    // A Negated{Joined{}} nodes needs to be mapped to the opposite Joined{}
+    // node. Parents need to point to the new Joined{} node. The old Joined{}
+    // node needs to be kept if it had parents other than the Negated node.
+    std::unordered_set<NodeId> orphaned_join_nodes;
+    std::unordered_set<NodeId> orphaned_negate_nodes;
```
Since all the node ids are consecutive, a vector of bools might be more effective (maybe not worrying about performance wise)

esseivaju replied:
let me make sure everything works before using a vector of bool 😅

sethrj commented on src/orange/orangeinp/CsgTreeUtils.cc:
```diff
@@ -165,6 +171,273 @@ void simplify(CsgTree* tree, NodeId start)
+    // nodes.
+    // TODO: extract the lambda to a free function
+    auto add_new_negated_nodes = [&](NodeId node_id) {
+        auto impl = [&](auto const& self, NodeId node_id) -> void {
```
Recursive lambdas before C++23??? You're a madman 😂

sethrj commented on test/orange/orangeinp/CsgTreeUtils.test.cc:
```diff
@@ -385,6 +385,119 @@ TEST_F(CsgTreeUtilsTest, calc_surfaces)
+    auto j0 = this->insert(Joined{op_and, {s0, n0}});
+    auto simplified = transform_negated_joins(tree_);
+
+    // check well-formed tree
```
For consistency with the rest of the code, please try to capitalize the comments and write them as a sentence without the trailing `.`, unless it's an adjective phrase describing data or something
```suggestion
    // Check a well-formed tree
```

esseivaju committed:
yet another yet another test case with Negates{Joined{Negated{Joined}}}

sethrj reviewed (changes requested):
Great commenting of what's going on! I know it's something of a work in progress, but a little restructuring would make this easier to review 😅 I think making the big function into a detail class (even a single-header one) will make it easier to read and process; that'll also make it easier to design a "visitor". Also don't forget the `Overload` class in `corecel/cont/VariantUtils.hh` as an alternative to writing a whole standalone class.

esseivaju commented:
> Great commenting of what's going on! I know it's something of a work in progress, but a little restructuring would make this easier to review 😅 I think making the big function into a detail class (even a single-header one) will make it easier to read and process; that'll also make it easier to design a "visitor". Also don't forget the `Overload` class in `corecel/cont/VariantUtils.hh` as an alternative to writing a whole standalone class.

I had to add comments or even I'd get lost during the implementation. This function is getting unwieldy and way too long as you mentioned, thanks largely to the rough variant interface...
I wanted to get a few more complex test cases working first, now that it's done it might be worth double-checking that the simplification is the logic you'd expect and that I didn't make any mistakes in the test strings. I had to get the pen and paper out for this, I think they are correct. 
Next, I'll refactor the function to make it  shorter and easier to parse.

esseivaju committed:
first pass in factorizing simplification algorithm

esseivaju committed:
move steering of the algorithm inside the simplifier

esseivaju committed:
refactor some functions to visitor

esseivaju committed:
add more visitor

esseivaju committed:
more refactoring

esseivaju committed:
comment style

esseivaju commented:
It's in a better state now. It could still use better comments and variable names but at least it's easier to review. I think the most important thing to check is the correctness and exhaustiveness of test cases.

esseivaju committed:
cleanup imports

sethrj reviewed (commented):
One thing I see is that the CSG nodes don't map directly from old to new (naturally), but regions/volumes/etc. in the CSG unit that we're simplifying reference the original node numbers. We'll need to return not only a new CSG tree but also a vector of NodeIds where `remapped[old.get()] == new`

esseivaju commented:
Is it possible to initialize the regions/volumes/etc after we simplify the tree so that they directly use the correct node_id? If an arbitrary node can be referenced by an external component, there's no guarantee it still is in the simplified tree. 
For example, if we have a `Negated{Joined{}}` and that joined node has no other parents, it will be converted to the opposite join. If someone was keeping a node_id pointing directly to that joined node, there is no equivalent of the original joined node in the simplified tree.

esseivaju committed:
reorder functions

esseivaju committed:
structured bindings

sethrj commented:
@esseivaju I should've thought to mention this before, but unfortunately it's not going to be that easy: consider the following, a union of two spheres inside a volume enclosed by a larger sphere:
```mermaid
flowchart TB
  n00["t"]
  n01["~"]
  n01 --> n00
  n02["[EXTERIOR]\nbound@s"]
  n03["bound\n~"]
  n03 --> n02
  n04["bottomsph@s"]
  n05["bottomsph\n~"]
  n05 --> n04
  n06["topsph@s"]
  n07["topsph\n~"]
  n07 --> n06
  n08["union\n|"]
  n08 --> n05
  n08 --> n07
  n09["~"]
  n09 --> n08
  n10["fill\n&"]
  n10 --> n03
  n10 --> n09
subgraph Volumes
  v02(["[EXTERIOR]@global"])
  v08(["union_vol"])
  v10(["fill_vol"])
end
  v02 <--> n02
  v08 <--> n08
  v10 <--> n10
```
We need to keep a reference both to `fill_vol` and `union_vol`.

Is the easy solution just to replace `~join(op, ...)` with `join(other_op, ~...)` on insertion to the CSG tree (which would allow us to skirt the "no backward substitution" requirement by injecting negated daughter nodes)? I was hoping we could do this with a "simplification" step but perhaps adding this to `CsgTree::insert` is the best way...

esseivaju committed:
add new test case and fix bug when handling double negation

esseivaju commented:
I don't think moving that to `insert` would solve the problem. In your example, it's impossible to keep the union node and have an equivalent tree after simplification even if you do it during insertion, you'd create a new root. The `intersect_vol` volume would have to point to the negated union. If the graph below is valid after simplification, then it's easy to provide a mapping of old volume to new (negated) volume.

```mermaid
flowchart TB
  n00["t"]
  n01["~"]
  n01 --> n00
  n02["[EXTERIOR]\nbound@s"]
  n03["bound\n~"]
  n03 --> n02
  n04["bottomsph@s"]
  n06["topsph@s"]
  n08["intersect\n&"]
  n08 --> n04
  n08 --> n06
  n10["fill\n&"]
  n10 --> n03
  n10 --> n08
subgraph Volumes
  v02(["[EXTERIOR]@global"])
  v08(["intersect_vol"])
  v10(["fill_vol"])
end
  v02 <--> n02
  v08 <--> n08
  v10 <--> n10
```

esseivaju commented:
@sethrj where is the Volume<->NodeId mapping stored?

sethrj commented:
@esseivaju These aren't internally saved anywhere in the CSG tree, but there's a vector of volumes in `CsgUnit` at `orange/orangeinp/detail/CsgUnit.hh`:
```c++
    std::vector<NodeId> volumes;  //!< CSG node of each volume
```
These will all be final at the time when we simplify the tree.

esseivaju commented:
And is this not going to pose an issue to have that extra disjoint tree when the tree is converted to a logical expression? Maybe we'll need to select which roots we want to convert.

sethrj commented:
@esseivaju Yeah. We're creating the logical trees from *each* of the volumes as a root. Lots of the time, a given volume doesn't touch every node in the tree.

esseivaju committed:
simplify return statement

esseivaju committed:
move volumes vector to CsgTree

esseivaju committed:
keep nodes matching to a volume and update volume nodes in the simplified tree

esseivaju committed:
check that the volume points to a valid node id

esseivaju committed:
inline in declaration

esseivaju committed:
refactor loop when checking volume roots

esseivaju committed:
simple test case checking that volumes are kept

esseivaju committed:
Merge branch 'develop' into simplify-negated-join

esseivaju committed:
cast

esseivaju commented:
Okay, I've moved the `volumes` to the `CsgTree` and updated the simplification example. I'll need to add more complex test cases. This now simplifies to the following tree:

```mermaid
flowchart TB
  n00["t"]
  n01["~"]
  n01 --> n00
  n02["topsph@s"]
  n03["bottomsph@s"]
  n04["topsph\n~"]
  n04 --> n02
  n05["bottomsph\n~"]
  n05 --> n03
  n06["intersect\n&"]
  n06 --> n02
  n06 --> n03
  n07["union\n|"]
  n07 --> n04
  n07 --> n05
  n08["[EXTERIOR]\nbound@s"]
  n09["bound\n~"]
  n09 --> n08
  n10["fill\n&"]
  n10 --> n06
  n10 --> n09
subgraph Volumes
  v02(["[EXTERIOR]@global"])
  v08(["union_vol"])
  v10(["fill_vol"])
end
  v02 <--> n08
  v08 <--> n07
  v10 <--> n10
```

esseivaju committed:
add top-level volume

esseivaju committed:
also check for volumes pointing to negated Join

sethrj commented on src/orange/orangeinp/CsgTree.hh:
```diff
@@ -77,6 +77,12 @@ class CsgTree
+    inline void insert_volume(NodeId);
+
+    //! List of volumes defined in this tree
+    std::vector<NodeId> const& volumes() const { return volumes_; }
```
Maybe add a type alias

sethrj commented on src/orange/orangeinp/UnitProto.cc:
```diff
@@ -124,7 +124,7 @@ void UnitProto::build(ProtoBuilder& input) const
     // Save unit's bounding box
     {
-        NodeId node_id = csg_unit.volumes[orange_exterior_volume.get()];
+        NodeId node_id = csg_unit.tree.volumes()[orange_exterior_volume.get()];
```
Keeping a local reference `auto const& volumes = csg_unit.tree.volumes()` would make this function cleaner

sethrj commented on src/orange/orangeinp/CsgTreeIO.json.cc:
```diff
@@ -129,11 +129,12 @@ void to_json(nlohmann::json& j, CsgUnit const& unit)
 
     j["volumes"] = [&unit] {
-        CELER_ASSERT(unit.volumes.size() == unit.fills.size());
+        CELER_ASSERT(unit.tree.volumes().size() == unit.fills.size());
```
I think we should also annotate the volumes in the CSG tree as part of its output, now that it actually contains the volumes.

esseivaju replied:
Do you mean that you'd have 
```json
{
"tree:" { "nodes": [], "volumes": []}
}
```
instead of the "tree" key being a list?

sethrj replied:
Either that, or add an additional component to the volume nodes: 
```json
["t",["~",0],["S",0],["S",1],["~",3,0],["S",2],["~",5],["&",[2,4,6], 1],["S",3],["~",8],["&",[2,4,9]],["~",7],["&",[10,11]],["S",4],["~",2],["|",[13,14]]]
```
but that's dumb, I think replacing the list with a CSG tree would be more sensible. Then we could even add a `_format` key...

esseivaju replied:
mmh, an entry for a volume needs material data from the CsgUnit which we can't pass to the `to_json` serialization function

esseivaju replied:
With the following changes, many tests fail because of the new structure. This would require many changes. It may be better to do that in a separate PR.
```diff
diff --git a/src/orange/orangeinp/CsgTreeIO.json.cc b/src/orange/orangeinp/CsgTreeIO.json.cc
index f355328c1..f3ec75bf1 100644
--- a/src/orange/orangeinp/CsgTreeIO.json.cc
+++ b/src/orange/orangeinp/CsgTreeIO.json.cc
@@ -65,11 +65,21 @@ void to_json(json& j, CsgTree const& tree)
 {
     ContainerVisitor<CsgTree const&, CsgTree::NodeId> visit_tree{tree};
 
-    j = json::array();
+    j = json::object();
+
+    json nodes = json::array();
     for (auto n : range(CsgTree::NodeId{tree.size()}))
     {
-        j.push_back(visit_tree(JsonConverter{}, n));
+        nodes.push_back(visit_tree(JsonConverter{}, n));
+    }
+    j["nodes"] = nodes;
+
+    json volumes = json::array();
+    for (auto id : tree.volumes())
+    {
+        volumes.push_back(id.get());
     }
+    j["volumes"] = volumes;
 }
 
 namespace detail
@@ -128,18 +138,19 @@ void to_json(nlohmann::json& j, CsgUnit const& unit)
         return result;
     }();
 
-    j["volumes"] = [&unit] {
-        CELER_ASSERT(unit.tree.volumes().size() == unit.fills.size());
-        json result = json::array();
-        for (auto i : range(unit.tree.volumes().size()))
+    j["fills"] = [&volumes = unit.tree.volumes(), &fills = unit.fills] {
+
+        CELER_ASSERT(volumes.size() == fills.size());
+        json result = json::object();
+        for (auto const& i : range(fills.size()))
         {
-            auto entry
-                = nlohmann::json{{"csg_node", unit.tree.volumes()[i].get()}};
-            if (auto* m = std::get_if<GeoMaterialId>(&unit.fills[i]))
+
+            auto entry = json::object();
+            if (auto* m = std::get_if<GeoMaterialId>(&fills[i]))
             {
                 entry["material"] = m->unchecked_get();
             }
-            else if (auto* d = std::get_if<Daughter>(&unit.fills[i]))
+            else if (auto* d = std::get_if<Daughter>(&fills[i]))
             {
                 entry["universe"] = d->universe_id.unchecked_get();
                 entry["transform"] = d->transform_id.unchecked_get();
@@ -149,15 +160,6 @@ void to_json(nlohmann::json& j, CsgUnit const& unit)
         return result;
     }();
 
-    j["fills"] = [&volumes = unit.tree.volumes()] {
-        json result = json::array();
-        for (auto const& v : volumes)
-        {
-            result.push_back(v.unchecked_get());
-        }
-        return result;
-    }();
-
     if (unit.background)
     {
         j["background"] = unit.background.get();
```

esseivaju committed:
alias

esseivaju committed:
coalesce sets mapping old to new node ids

esseivaju committed:
coalesce sets mapping orphanes nodes

esseivaju committed:
move implementation to cc file

esseivaju committed:
monadic-style operations on helper struct

esseivaju committed:
use extract from set and map

esseivaju committed:
documentation and renaming

esseivaju committed:
more doc

esseivaju committed:
simplify check when inserting a join node

esseivaju committed:
doc and typo

esseivaju committed:
use vector of bool

esseivaju committed:
cleanup

esseivaju committed:
more typos

esseivaju committed:
a few more typos

esseivaju committed:
handle adding negated join with negated children

esseivaju committed:
remove unmod_mod and be more explicit in situations where we expect to find a modified node

esseivaju committed:
document when MatchingNodes::modified is set

esseivaju committed:
remove unnecessary insertion of a negated node

esseivaju committed:
simplify record_parents_for

esseivaju commented:
@sethrj I think this is ready for another look. The only thing I'm not convinced of yet is that `MatchingNodes::modified` is a bit overloaded and factorizing it into 3 variables would make the algorithm easier to follow.

sethrj commented on src/orange/orangeinp/CsgTree.hh:
```diff
@@ -114,6 +123,16 @@ auto CsgTree::insert(LocalSurfaceId s) -> Insertion
+/*!
+ * Insert a new volume which root is the node id.
+ */
+inline void CsgTree::insert_volume(NodeId node_id)
```
Redundant
```suggestion
void CsgTree::insert_volume(NodeId node_id)
```

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.hh:
```diff
@@ -0,0 +1,124 @@
+    struct MatchingNodes
+    {
+        //! Set if a node has the exact same node in the simplified tree
+        std::optional<NodeId> unmodified;
```
Since an unset `NodeId` is false (it's supposed to act rather  like "optional") is redundant: you should be able to just use `NodeId unmodified`, same with `modified`

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+CsgTree DeMorganSimplifier::operator()()
+{
+    find_join_negations();
+    return build_simplified_tree();
```
Use `this->member_func()` for improved readability (and to make it more obvious why the lambdas need to capture `this`

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+                    record_parent_for(join_operand);
+                }
+                // and the Joined node itself needs to be kept
+                CELER_EXPECT(std::holds_alternative<Joined>(tree_[node_id]));
```
This should be completely redundant with the `std::visit` call, right? Since there's no precondition that could trigger the assertion it should be `CELER_ASSERT`. And really, since there's no programming error you could make that would trigger this, you should probably delete it...

esseivaju replied:
Right, it is redundant I'll just remove that

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+    // Consider a "virtual" parent for these nodes
+    std::unordered_set<NodeId> volume_roots{tree_.volumes().cbegin(),
+                                            tree_.volumes().cend()};
+    for (auto node_id : volume_roots)
```
No two volumes should point to the same node: that implies two volumes occupy the same exact region of space, which is a geometry error. We could in theory check this in the CSG tree itself. For now, I think the code will still work even if the same NodeId is shared by two volumes...
```suggestion
    for (auto node_id : tree_.volumes())
```

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+{
+    CELER_EXPECT(negated_id < tree_.size());
+
+    CELER_ASSERT(std::holds_alternative<Negated>(tree_[negated_id]));
```
```suggestion
    CELER_EXPECT(std::holds_alternative<Negated>(tree_[negated_id]));
```
This one is a precondition on input :) You can also use `CELER_ASSUME` to help the compiler optimize when assertions are disabled.

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+    CELER_ASSERT(std::holds_alternative<Joined>(tree_[negated_node->node]));
+
+    negated_join_nodes_[negated_node->node.get()] = true;
+    add_negation_for_operands(negated_node->node);
```
With `CELER_ASSUME`, a compiler with optimizations won't even generate the exception handling code for `std::get`, so you should be able to write this more cleanly as
```suggestion
    simplified_negated_nodes_[negated_id.get()] = true;

    NodeId negated_join = std::get<Negated>(tree_[negated_id]).node_id;
    CELER_ASSUME(std::holds_alternative<Joined>(tree_[negated_join]));
    negated_join_nodes_[negated_join.get()] = true;
    this->add_negation_for_operands(negated_join);
```

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+
+    for (auto const& join_operand : join_node->nodes)
+    {
+        std::visit(Overload{[&](Joined const&) {
```
I feel like a couple of chained `get_if`s may be more semantically appropriate for this case.

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+            Overload{
+                [&](Negated& negated) {
+                    negated.node
+                        = node_ids_translation_[negated.node].mod_unmod_or(
```
Maybe `replaced` would be be a better name for this function?

esseivaju replied:
it's not always replaced, e.g. if the child was a surface.

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+ */
+CsgTree DeMorganSimplifier::operator()()
+{
+    find_join_negations();
```
The ordering of the functions in this header is quite hard to follow... maybe order them in roughly the order that they could be called (sorted topological call graph)?
- construct
- operator()
- find join negations
- record_join_negation
- add negation for operand
- ....
- build simplified tree
- process_negated_joined_nodes

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+                // children
+                if (orphaned_nodes_[node_id.get()])
+                {
+                    CELER_EXPECT(
+                        std::holds_alternative<Negated>(tree_[node_id]));
+                    auto* orphan_node = std::get_if<Negated>(&tree_[node_id]);
```
Here you could just just use the argument to the lambda function right?

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,443 @@
+                // already inserted the opposite Joined node
+                if (simplified_negated_nodes_[node_id.get()])
+                {
+                    CELER_EXPECT(
+                        std::holds_alternative<Negated>(tree_[node_id]));
+
+                    auto& negated_node_child
+                        = std::get_if<Negated>(&tree_[node_id])->node;
```
same:
```suggestion
                    auto& negated_node_child = negated.node;
```

sethrj commented on test/orange/orangeinp/CsgTreeUtils.test.cc:
```diff
@@ -385,6 +385,306 @@ TEST_F(CsgTreeUtilsTest, calc_surfaces)
+        to_string(simplified));
+    // Check that the new volumes map to the correct node
+    auto& volumes = simplified.volumes();
+    EXPECT_EQ(volumes.size(), 3);
```
Good to use this form to prevent the code crashing in case of an error...
```suggestion
    ASSERT_EQ(volumes.size(), 3);
```

sethrj reviewed (commented):
This is an impressive piece of work. Could you please address a couple of these comments which hopefully will continue to make it easier to review? Can you also post another couple of before-and-after CSG trees so we can visually verify?

esseivaju committed:
address part of the comments

esseivaju committed:
reorder functions

esseivaju committed:
std::get_if instead of std::visit

esseivaju committed:
Merge branch 'develop' into simplify-negated-join

esseivaju commented:
This is the most complicated test problem, the ATLAS tilecal cylinder segment. Roots 16 and 27 are preserved so we have to duplicate many nodes.

Before simplification ([mermaidchart](https://www.mermaidchart.com/app/projects/98e19971-82da-477e-ab3b-dac122f9f53f/diagrams/9b72871c-2586-445a-bf7d-c696326fd8ac/version/v0.1/edit)):

```mermaid
flowchart TB
  n00["00:t"]
  n01["01:~"]
  n01 --> n00
  n02["02:S0"]
  n03["03:S1"]
  n04["04:~"]
  n04 --> n03
  n05["05:S2"]
  n06["06:~"]
  n06 --> n05
  n07["07:&"]
  n07 --> n02
  n07 --> n04
  n07 --> n06
  n08["08:S3"]
  n09["09:~"]
  n09 --> n08
  n10["10:&"]
  n10 --> n02
  n10 --> n04
  n10 --> n09
  n11["11:~"]
  n11 --> n10
  n12["12:S4"]
  n13["13:S5"]
  n14["14:&"]
  n14 --> n12
  n14 --> n13
  n15["15:&"]
  n15 --> n07
  n15 --> n11
  n15 --> n14
  n16["16:~"]
  n16 --> n15
  n17["17:S6"]
  n18["18:S7"]
  n19["19:~"]
  n19 --> n18
  n20["20:&"]
  n20 --> n06
  n20 --> n17
  n20 --> n19
  n21["21:&"]
  n21 --> n09
  n21 --> n17
  n21 --> n19
  n22["22:~"]
  n22 --> n21
  n23["23:S8"]
  n24["24:S9"]
  n25["25:&"]
  n25 --> n23
  n25 --> n24
  n26["26:&"]
  n26 --> n20
  n26 --> n22
  n26 --> n25
  n27["27:~"]
  n27 --> n26
  n28["28:&"]
  n28 --> n15
  n28 --> n27
```
after simplification ([mermaidchart](https://www.mermaidchart.com/app/projects/98e19971-82da-477e-ab3b-dac122f9f53f/diagrams/3cbc8f5c-863a-4fe5-bb27-3923c29f6b90/version/v0.1/edit)):

```mermaid
flowchart TB
  n00["00:t"]
  n01["01:~"]
  n01 --> n00
  n02["02:S0"]
  n03["03:~"]
  n03 --> n02
  n04["04:S1"]
  n05["05:~"]
  n05 --> n04
  n06["06:S2"]
  n07["07:~"]
  n07 --> n06
  n08["08:|"]
  n08 --> n03
  n08 --> n04
  n08 --> n06
  n09["09:&"]
  n09 --> n02
  n09 --> n05
  n09 --> n07
  n10["10:S3"]
  n11["11:~"]
  n11 --> n10
  n12["12:|"]
  n12 --> n03
  n12 --> n04
  n12 --> n10
  n13["13:&"]
  n13 --> n02
  n13 --> n05
  n13 --> n11
  n14["14:S4"]
  n15["15:~"]
  n15 --> n14
  n16["16:S5"]
  n17["17:~"]
  n17 --> n16
  n18["18:|"]
  n18 --> n15
  n18 --> n17
  n19["19:&"]
  n19 --> n14
  n19 --> n16
  n20["20:|"]
  n20 --> n08
  n20 --> n13
  n20 --> n18
  n21["21:&"]
  n21 --> n09
  n21 --> n12
  n21 --> n19
  n22["22:S6"]
  n23["23:~"]
  n23 --> n22
  n24["24:S7"]
  n25["25:~"]
  n25 --> n24
  n26["26:|"]
  n26 --> n06
  n26 --> n23
  n26 --> n24
  n27["27:|"]
  n27 --> n10
  n27 --> n23
  n27 --> n24
  n28["28:&"]
  n28 --> n11
  n28 --> n22
  n28 --> n25
  n29["29:S8"]
  n30["30:~"]
  n30 --> n29
  n31["31:S9"]
  n32["32:~"]
  n32 --> n31
  n33["33:|"]
  n33 --> n30
  n33 --> n32
  n34["34:|"]
  n34 --> n26
  n34 --> n28
  n34 --> n33
  n35["35:&"]
  n35 --> n21
  n35 --> n34
```

esseivaju commented:
And a smaller test case with volumes: 

Before:
```mermaid
flowchart TB
  n00["00:t"]
  n01["01:~"]
  n01 --> n00
  n02["02:S0"]
  n03["03:S1"]
  n04["04:~"]
  n04 --> n02
  n05["05:~"]
  n05 --> n03
  n06["06:|"]
  n06 --> n04
  n06 --> n05
  n07["07:~"]
  n07 --> n06
  n08["08:S2"]
  n09["09:~"]
  n09 --> n08
  n10["10:&"]
  n10 --> n07
  n10 --> n09
  subgraph Volumes
    v0["vol0"]
    v1["vol1"]
    v2["vol2"]
  end
  v0 --> n06
  v1 --> n10
  v2 --> n07
```

After simplification:

```mermaid
flowchart TB
  n00["00:t"]
  n01["01:~"]
  n01 --> n00
  n02["02:S0"]
  n03["03:S1"]
  n04["04:~"]
  n04 --> n02
  n05["05:~"]
  n05 --> n03
  n06["06:&"]
  n06 --> n02
  n06 --> n03
  n07["07:|"]
  n07 --> n04
  n07 --> n05
  n08["08:S2"]
  n09["09:~"]
  n09 --> n08
  n10["10:&"]
  n10 --> n06
  n10 --> n09
  subgraph Volumes
    v0["vol0"]
    v1["vol1"]
    v2["vol2"]
  end
  v0 --> n07
  v1 --> n10
  v2 --> n06
```
Without volumes, this tree simplifies to:
```mermaid
flowchart TB
  n00["00:t"]
  n01["01:~"]
  n01 --> n00
  n02["02:S0"]
  n03["03:S1"]
  n04["04:&"]
  n04 --> n02
  n04 --> n03
  n05["05:S2"]
  n06["06:~"]
  n06 --> n05
  n07["07:&"]
  n07 --> n04
  n07 --> n06
```

esseivaju committed:
test to fix

sethrj commented:
One thing I notice (not to do with your work or in scope for this PR) is that we have `join(and, [join(and, [..]), ...])` which should probably be "canonicalized" into a single `join`. I wonder if that can easily be implemented by `NodeSimplifier`...

(EDIT: Done in #1784)

esseivaju commented:
> I wonder if that can easily be implemented by NodeSimplifier...

If you know that there are no other parents (volume) for that nested join, then I suppose that the NodeSimplifier should be able to do that.

esseivaju committed:
record parents during the first pass instead of tagging nodes as orphans and don't cache negated nodes with joined child

esseivaju committed:
some cleanup

esseivaju committed:
minor simplification

esseivaju committed:
more simplification

esseivaju committed:
update expects

esseivaju commented:
@sethrj I reworked the code to remove `MatchingNodes::modified` and use multiple variables instead, being more explicit about which value is needed in a given situation. I think this makes the algorithm clearer.

esseivaju committed:
use vector instead of map

esseivaju committed:
format

esseivaju committed:
use matrix of bool instead of vector of set for parents

esseivaju committed:
remove if

esseivaju committed:
remove unnecessary double_negation indirection

esseivaju committed:
utility to lookup an equal node in the simplified tree

esseivaju committed:
create a rudimentary mdspan view of the vector<bool>

esseivaju committed:
constexpr

esseivaju committed:
unused declaration

esseivaju committed:
fixup

esseivaju committed:
update documentation

esseivaju committed:
group expect

esseivaju commented:
@sethrj I'm finally happy with the implementation so it should be good for review.

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.hh:
```diff
@@ -0,0 +1,143 @@
+        NodeId unmodified;
+
+        //! Indirections to new simplified join following DeMorgan's law
+        //! Set iif the original node is a negated node.
```
Is this `iff` ("if and only if") or a typo?

esseivaju replied:
typo

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.hh:
```diff
@@ -0,0 +1,143 @@
+        NodeId equivalent_node() const;
+    };
+
+    //! Rudimentary 2D square matrix view of a vector<bool>
```
Should the matrix be square? I thought the number of nodes was likely to change. Also it means that the storage requirements (and cost?) scale quadratically with the number of CSG nodes... which could be unpleasant for some of the geometries. (A fair number have hundreds of volumes each of which has  a couple of faces.)

esseivaju replied:
This is only used to represent the parent relationship in the original `CsgTree`so the number of nodes never changes. Storage does scale quadratically but at least vector<bool> is very compact. Under a certain number of CsgNode per tree, a sparse matrix could easily use more space since you'd need to store row/col indices though it would still be better to use it. The quadratic memory scaling starts to become significant around 1000 CsgNode (100Mb for that matrix of parents)

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.hh:
```diff
@@ -0,0 +1,143 @@
+    //! an opposite join node with negated operands
+    std::vector<bool> negated_join_nodes_;
+
+    //! Parents matrix. For nodes n1, n2, if n1 * tree_.size() + n2 is set, it
+    //! means that n2 is a parent of n1
+    Matrix2D parents_;
```
Yeah I really think this should be sparse...

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.hh:
```diff
@@ -0,0 +1,143 @@
+    //! CsgTree node 1 is always a Negated node parent of node 0, so we can
+    //! reuse that bit to tell if a node has a parent as it's never set for
+    //! node id >= 2
+    static constexpr auto has_parents_index{NodeId{1}};
```
Even though these are constexpr, since they act like private member data should we give them underscores as well? When I saw the constant used I wasn't sure if it was a local variable, global constant, etc.

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,449 @@
+                    return true;
+                }
+
+                //
```
```suggestion
```

sethrj commented on src/orange/orangeinp/detail/DeMorganSimplifier.cc:
```diff
@@ -0,0 +1,449 @@
+
+                // Insert the opposite join
+                auto const& [op, nodes] = joined;
+                std::vector<NodeId> operands;
+                operands.reserve(nodes.size());
+
+                // Negate each operand, pointing to node ids in the simplified
+                // tree
+                for (auto n : nodes)
+                {
+                    // Negation of a negated operand cancel each other, we can
+                    // just use the child of that negated operand
+                    if (auto* neg = std::get_if<Negated>(&tree_[n]))
+                    {
+                        // We should have recorded that this node was necessary
+                        // for a join
+                        CELER_EXPECT(
+                            node_ids_translation_[neg->node.get()].unmodified);
+                        operands.push_back(
+                            node_ids_translation_[neg->node.get()].unmodified);
+                    }
+                    else
+                    {
+                        // Otherwise, we should have inserted a negated
+                        // version of that operand in the simplified tree.
+                        // It's either a simplified join or a negated node
+                        operands.push_back([&] {
+                            auto& trans = node_ids_translation_[n.get()];
+                            CELER_EXPECT(trans.new_negation
+                                         || trans.opposite_join);
+                            if (trans.new_negation)
+                                return trans.new_negation;
+                            return trans.opposite_join;
+                        }());
+                    }
+                }
```
This looks like a good candidate for another helper member function
```c++
std::vector<NodeId> DeMorganSimplifier::build_negated_nodes(std::vector<NodeId> const& nodes) const
```

sethrj reviewed (approved):
Great refactor @esseivaju , sorry it took me two weeks to review! We can postpone the "2D matrix" reconsideration to a follow-on PR since that's an optimization. I think we should (for testing) activate the simplifier as part of the UnitProto construction and compare before/after build times and memory usage for our suite of benchmark problems (including the super big ugly nasty ones).

esseivaju committed:
address comments

esseivaju committed:
Merge branch 'develop' into simplify-negated-join
